### PR TITLE
improved ux in attribute section by assuming all attributes are configured when no attributes are configured for oidc apps

### PR DESCRIPTION
**Pull request by mpmadhavig**
### Proposed changes in this pull request
As mentioned in this [issue](https://github.com/wso2-enterprise/asgardeo-product/issues/5443), when a new OIDC app is created, the developer need to configure attributes one by one as requested using the console, in order to access them from the client application.  

- This fix will improve the UX of the developer by assuming all attributes are allowed when no attributes are configured by the developer.

### Related Issues

- https://github.com/wso2-enterprise/asgardeo-product/issues/5443

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java [hunk 1]
```diff
@@ -72,6 +72,9 @@ public class ServiceProvider implements Serializable {
     @XmlTransient
     private String tenantDomain;
 
+    @XmlTransient
+    private boolean isIllegibleForDefaultClaimMappings;
```
Comment (mefarazath) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java:76]: Would be good to think of a simple name that conveys the purpose?

Comment (mefarazath) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java:76]: Also shouldn't this go inside claim configuration model?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 2]
```diff
@@ -3296,4 +3296,47 @@ public static String buildCallerPathRedirectURL(String callerPath, Authenticatio
...
+
+        } catch (CarbonException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Could not retrieve the realm", e);
+            }
+            return new ClaimMapping[0];
+        }
+
+
+
```
Comment (piyarathnalakmali) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:3340]: Remove unnecessary new lines.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java (hunk 1 above)
Comment (mpmadhavig) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java]: > 

Removed app level flag

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 3]
```diff
@@ -1,17 +1,17 @@
-/*
- * Copyright (c) 2013, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+/**
+ * Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```
Comment (piyarathnalakmali) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:2]: According to my understanding we don't need to update the year in the license header if only the created year is mentioned. Shall we confirm this and update accordingly.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 4]
```diff
@@ -129,22 +129,29 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
-                getClaimMappings();
-
-        if (spClaimMappings == null) {
-            spClaimMappings = new HashMap<>();
-        }
+        boolean useDefaultClaimMappings = canUseDefaultClaimMappings(appConfig, context);
+        Map<String, String> spClaimMappings = getSpToLocalClaimMappings(useDefaultClaimMappings, appConfig,
+                claimManager);
+        Map<String, String> spRequestedClaimMappings = getRequestedClaimMappings(useDefaultClaimMappings, appConfig,
+                spClaimMappings);
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:151]: add useDefaultClaimMappings variable initialization inside the pvt methods.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 5]
```diff
@@ -129,22 +129,29 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
 
-        ClaimMapping[] idPClaimMappings = context.getExternalIdP().getClaimMappings();
+        ApplicationConfig appConfig = context.getSequenceConfig().getApplicationConfig();
+        AuthenticatedUser authenticatedUser = getAuthenticatedUser(stepConfig, context);
+        String tenantDomain = authenticatedUser.getTenantDomain();
+        UserRealm realm = getUserRealm(tenantDomain);
+        if (realm == null) {
+            log.warn("No valid tenant domain provider. No claims returned back");
+            return new HashMap<>();
+        }
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:139]: check exception handling can replace this.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 6]
```diff
@@ -129,22 +129,29 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
             throws FrameworkException {
 
-        ClaimMapping[] idPClaimMappings = context.getExternalIdP().getClaimMappings();
+        ApplicationConfig appConfig = context.getSequenceConfig().getApplicationConfig();
+        AuthenticatedUser authenticatedUser = getAuthenticatedUser(stepConfig, context);
+        String tenantDomain = authenticatedUser.getTenantDomain();
+        UserRealm realm = getUserRealm(tenantDomain);
+        if (realm == null) {
+            log.warn("No valid tenant domain provider. No claims returned back");
+            return new HashMap<>();
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:138]: Use collections class and get empty map.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 7]
```diff
@@ -129,22 +129,29 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
                                                         StepConfig stepConfig, AuthenticationContext context)
             throws FrameworkException {
 
-        ClaimMapping[] idPClaimMappings = context.getExternalIdP().getClaimMappings();
+        ApplicationConfig appConfig = context.getSequenceConfig().getApplicationConfig();
+        AuthenticatedUser authenticatedUser = getAuthenticatedUser(stepConfig, context);
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:133]: check whether the null is handled?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 8]
```diff
@@ -536,6 +534,57 @@ protected Map<String, String> handleLocalClaims(String spStandardDialect,
...
+                    .getAllClaimMappings(ApplicationConstants.LOCAL_IDP_DEFAULT_CLAIM_DIALECT);
+
+            for (org.wso2.carbon.user.api.ClaimMapping mapping : claimMappings) {
+                String claimURI = mapping.getClaim().getClaimUri();
+                spToLocalClaimMappings.put(claimURI, claimURI);
+            }
+
+        } catch (UserStoreException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Could not retrieve the Claim Manager", e);
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:581]: add more usefull info

Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:581]: handle or throw

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 9]
```diff
@@ -536,6 +534,57 @@ protected Map<String, String> handleLocalClaims(String spStandardDialect,
...
+
+    private Map<String, String> getDefaultSpToLocalClaimMappings(ClaimManager claimManager) {
+
+        Map<String, String> spToLocalClaimMappings = new HashMap<>();
+        try {
+            org.wso2.carbon.user.api.ClaimMapping[] claimMappings = claimManager
+                    .getAllClaimMappings(ApplicationConstants.LOCAL_IDP_DEFAULT_CLAIM_DIALECT);
+
+            for (org.wso2.carbon.user.api.ClaimMapping mapping : claimMappings) {
+                String claimURI = mapping.getClaim().getClaimUri();
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:575]: check getclaim() return null

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/SSOConsentServiceImpl.java [hunk 10]
```diff
@@ -1,17 +1,19 @@
...
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/SSOConsentServiceImpl.java:16]: change only Inc to LLC

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 11]
```diff
@@ -536,6 +534,57 @@ protected Map<String, String> handleLocalClaims(String spStandardDialect,
...
+
+        if (useDefaultClaimMappings) {
+            return getDefaultRequestedClaimMappings(spToLocalClaimMappings);
+        }
+        return appConfig.getRequestedClaimMappings();
+    }
+
+    private boolean canUseDefaultClaimMappings(ApplicationConfig appConfig, AuthenticationContext context) {
+
+        boolean isClaimConfiguredForApp = !appConfig.getClaimMappings().isEmpty();
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:557]: remove protocol specific log

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 12]
```diff
@@ -536,6 +534,57 @@ protected Map<String, String> handleLocalClaims(String spStandardDialect,
...
+        boolean isOIDCRequest = context.getRequestType().equals(FrameworkConstants.RequestType.CLAIM_TYPE_OIDC);
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:558]: remove protocol specific logic

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/AbstractFrameworkTest.java [hunk 13]
```diff
@@ -43,6 +44,7 @@ protected AuthenticationContext getAuthenticationContext(ServiceProvider service
         authenticationContext.setTenantDomain("test_domain");
         authenticationContext.setCurrentStep(1);
         authenticationContext.setContextIdentifier(UUID.randomUUID().toString());
+        authenticationContext.setRequestType(FrameworkConstants.RequestType.CLAIM_TYPE_OIDC);
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/AbstractFrameworkTest.java:47]: remove protocol specific logic 

add a new test to check the not configured claims configuration

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 14]
```diff
@@ -3296,4 +3296,44 @@ public static String buildCallerPathRedirectURL(String callerPath, Authenticatio
...
+            org.wso2.carbon.user.api.ClaimMapping[] claimMappings = claimManager
+                    .getAllClaimMappings(ApplicationConstants.LOCAL_IDP_DEFAULT_CLAIM_DIALECT);
+
+            for (org.wso2.carbon.user.api.ClaimMapping mapping : claimMappings) {
+                String claimURI = mapping.getClaim().getClaimUri();
+                ClaimMapping claim = ClaimMapping.build(claimURI, claimURI, null, true,
+                        false);
+                defaultClaimMappings.add(claim);
+            }
```
Comment (mpmadhavig) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:3323]: try to use existing functions

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 3 above)
Comment (mpmadhavig) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java]: resolved with https://github.com/wso2/carbon-identity-framework/pull/4244/commits/9a47021d8fafc7687f29ea4cf5cc08b7ee34953d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 12 above)
Comment (mpmadhavig) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java]: fixed with https://github.com/wso2/carbon-identity-framework/pull/4244/commits/4f2be12a856e0ec20b5c0fc918707a7bdc4592c0

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 4 above)
Comment (mpmadhavig) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java]: fixed with https://github.com/wso2/carbon-identity-framework/pull/4244/commits/4f2be12a856e0ec20b5c0fc918707a7bdc4592c0

**Comment by mpmadhavig** [comment]
The path of this issue has changed and hence closing the PR.

**Pull request closed without merging**